File: frontend/features/chat/types/api/session.ts
/**
 * Session-related API types matching backend schemas
 */

import type { ApiStatePatch } from "./callback";

export interface SessionCreateRequest {
  config?: TaskConfig | null;
}

export interface SessionUpdateRequest {
  status?: string | null;
  sdk_session_id?: string | null;
  title?: string | null;
  workspace_archive_url?: string | null;
  state_patch?: ApiStatePatch | null;
  workspace_files_prefix?: string | null;
  workspace_manifest_key?: string | null;
  workspace_archive_key?: string | null;
  workspace_export_status?: string | null;
  project_id?: string | null;
}

export interface SessionCancelRequest {
  reason?: string | null;
}

export interface SessionCancelResponse {
  session_id: string;
  status: string;
  canceled_runs: number;
  expired_user_input_requests: number;
  executor_cancelled: boolean;
}

export interface SessionResponse {
  session_id: string; // UUID
  user_id: string;
  sdk_session_id: string | null;
  project_id?: string | null;
  title?: string | null;
  config_snapshot: Record<string, unknown> | null;
  workspace_archive_url: string | null;
  state_patch?: ApiStatePatch | null;
  workspace_export_status?: string | null;
  workspace_files_prefix?: string | null;
  workspace_manifest_key?: string | null;
  workspace_archive_key?: string | null;
  status: string;
  created_at: string; // ISO datetime
  updated_at: string; // ISO datetime
}

export interface SessionStateResponse {
  session_id: string;
  status: string;
  state_patch?: ApiStatePatch | null;
  workspace_export_status?: string | null;
  updated_at: string;
}

export interface MessageResponse {
  id: number;
  role: string;
  content: Record<string, unknown>;
  created_at: string; // ISO datetime
  updated_at: string; // ISO datetime
}

export interface ToolExecutionResponse {
  id: string; // UUID
  message_id: number | null;
  tool_use_id: string | null;
  tool_name: string;
  tool_input: Record<string, unknown> | null;
  tool_output: Record<string, unknown> | null;
  is_error: boolean;
  duration_ms: number | null;
  created_at: string; // ISO datetime
}

export interface UsageResponse {
  total_cost_usd: number | null;
  total_duration_ms: number | null;
  usage_json: Record<string, unknown> | null;
}

export interface InputFile {
  [x: string]: unknown;
  id?: string | null;
  type?: string;
  name: string;
  source: string;
  size?: number | null;
  content_type?: string | null;
  path?: string | null;
}

export interface TaskConfig {
  repo_url?: string | null;
  git_branch?: string; // defaults to "main"
  /** Env var key holding a GitHub token (e.g. "GITHUB_TOKEN"). */
  git_token_env_key?: string | null;
  /** Built-in browser capability toggle (Playwright MCP is injected internally). */
  browser_enabled?: boolean;
  /** MCP server enable/disable toggles (true=enabled, false=disabled).
   *  Servers not in this object use their default enabled state from user installations.
   */
  mcp_config?: Record<string, boolean>;
  /** Skill enable/disable toggles (true=enabled, false=disabled).
   *  Skills not in this object use their default enabled state from user installs.
   */
  skill_config?: Record<string, boolean>;
  /** Plugin enable/disable toggles (true=enabled, false=disabled).
   *  Plugins not in this object use their default enabled state from user installs.
   */
  plugin_config?: Record<string, boolean>;
  /** Optional explicit subagent selection (by id). */
  subagent_ids?: number[];
  input_files?: InputFile[];
}

File: backend/app/schemas/session.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field
from app.schemas.callback import AgentCurrentState
from app.schemas.input_file import InputFile


class TaskConfig(BaseModel):
    """Task configuration."""

    repo_url: str | None = None
    git_branch: str = "main"
    # Optional env var key holding a GitHub token (e.g. "GITHUB_TOKEN").
    git_token_env_key: str | None = None
    # Built-in browser capability toggle (Playwright MCP is injected internally by the executor).
    browser_enabled: bool = False
    # MCP server enable/disable toggles (true=enabled, false=disabled).
    # Servers not in this dict use their default enabled state from user installations.
    mcp_config: dict[str, bool] = Field(default_factory=dict)
    # Skill enable/disable toggles (true=enabled, false=disabled).
    # Skills not in this dict use their default enabled state from user installations.
    skill_config: dict[str, bool] = Field(default_factory=dict)
    # Optional explicit subagent selection (by id). When omitted, backend resolves
    # enabled subagents as defaults.
    subagent_ids: list[int] = Field(default_factory=list)
    input_files: list[InputFile] = Field(default_factory=list)


class SessionCreateRequest(BaseModel):
    """Request to create a session."""

    config: TaskConfig | None = None
    project_id: UUID | None = None


class SessionUpdateRequest(BaseModel):
    """Request to update a session."""

    status: str | None = None
    sdk_session_id: str | None = None
    title: str | None = None
    workspace_archive_url: str | None = None
    project_id: UUID | None = None
    state_patch: dict[str, Any] | None = None
    workspace_files_prefix: str | None = None
    workspace_manifest_key: str | None = None
    workspace_archive_key: str | None = None
    workspace_export_status: str | None = None


class SessionResponse(BaseModel):
    """Session response."""

    session_id: UUID = Field(validation_alias="id")
    user_id: str
    project_id: UUID | None
    sdk_session_id: str | None
    title: str | None = None
    config_snapshot: dict[str, Any] | None
    workspace_archive_url: str | None
    state_patch: AgentCurrentState | None = None
    workspace_export_status: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class SessionStateResponse(BaseModel):
    """Session state response."""

    session_id: UUID = Field(validation_alias="id")
    status: str
    state_patch: AgentCurrentState | None = None
    workspace_export_status: str | None = None
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class SessionCancelRequest(BaseModel):
    """Request to cancel a session (cancel all unfinished runs)."""

    reason: str | None = None


class SessionCancelResponse(BaseModel):
    """Session cancel response."""

    session_id: UUID
    status: str
    canceled_runs: int = 0
    expired_user_input_requests: int = 0
    executor_cancelled: bool = False
